### ISSUE TITLE
[NF] Success with Empty Body (204) / Ignored Success Body

### ISSUE THREAD

**Issue by lucianocn** [haroldadmin/NetworkResponseAdapter#9]
Hi @haroldadmin ,
First of all, amazing library!

I am working on a new feature to adapt it to my project and I believe can be useful for other people.

- API requests can return 204 (no content) with an empty body and should be treated as a successful response.
- Ignored body on success may occur in REST requests, just relying on the status code. However, I would still need to parse unsuccessful responses.

I intend to work on it in 1 week and I would like to know your concerns or any tips.

**Comment by haroldadmin** [comment]
Hi! Thanks for bringing up this issue. I did not think of your use-case when writing the library, but it's great that you're adding support for it. 

The changes would mainly be in the `DeferredNetworkResponseAdapter.kt` and `NetworkResponseCall.kt` files. The error handling code needs to be tweaked a little.

Please add tests for whatever new functionality you add, and let me know if you need any help with implementing this!

**Comment by haroldadmin** [comment]
@lucianocn Are you still working on this? If not, please let me know so I can pick it up.

**Comment by lucianocn** [comment]
> @lucianocn Are you still working on this? If not, please let me know so I can pick it up.

Unfortunately I still haven't had time to work on this feature. If possible, I will be very grateful.

**Comment by haroldadmin** [comment]
No worries, I'll try to make this change in a couple of days.

**Comment by haroldadmin** [comment]
@lucianocn The `NetworkResponse.Success` class contains a `body` parameter which is of a non-null type. An empty response is treated as a null response, which means that I can not assign it as the value to the `body` parameter.

A hacky workaround for this would be to use some reflection magic and set a null value on it, but I don't think it's a good choice to fight the type-system of Kotlin just to fulfill one edge case. Not sure what's the best way to proceed here, because I *really* don't want to make the `body` parameter type nullable.

I'll think more about a possible solution to this, but until then consider this issue iceboxed.

**Comment by lukaszkalnik** [comment]
I haven't looked into the source code yet, but would returning `Unit` be an option for 204 responses?

**Comment by lukaszkalnik** [comment]
Ok, I see it won't work as the return type is generic. Then the nullability would be the only option.
It would actually make sense, because what we're saying is "This is a generic CallAdapter for any response that can return an instance of the generic type `S`, including for responses which return no instance, hence `null`". Null in Java/Kotlin belongs to any type, so this is the only value which fulfills the generic type contract (`Unit` is not, as it only belongs to type `Unit`).

**Comment by haroldadmin** [comment]
> I haven't looked into the source code yet, but would returning `Unit` be an option for 204 responses? It's semantics in Kotlin is "value not relevant".

I like the idea of representing empty response bodies with `Unit`. It is possible to add a special case to the call adapter to return `Unit` in the case of a null response body, provided that the success type of the service method is also declared as Unit:

```kotlin
@GET("/empty-body-endpoint")
suspend fun getEmptyBodyResponse(): NetworkResponse<Unit, ErrorType>
```

What do you think @lucianocn @lukaszkalnik ?

**Comment by lukaszkalnik** [comment]
This looks good, assuming the service is guaranteed to always return the same kind of response from the same call (i.e. it will always return either some data or 204 No Content, but mixing the one with another for one call is not allowed). Then the API user can declare some methods as returning `Unit` and others as returning his expected domain model datatype.

**Comment by lucianocn** [comment]
Sounds good! I like it and I believe that `Unit` is the correct solution.

**Comment by haroldadmin** [comment]
Great. I'll try to incorporate this change in the next release. Should be out in a few days I guess.

**Comment by haroldadmin** [comment]
@lucianocn @lukaszkalnik These changes are added in PR #14. I'd appreciate any feedback that you might have on it. If things look good, I'll merge it and you can try it out in a SNAPSHOT build from Jitpack.

**Comment by lucianocn** [comment]
I guess the second point in the main comment is not handled by ResponseHandler.kt:
- Ignored body on success may occur in REST requests, just relying on the status code. However, I would still need to parse unsuccessful responses (In that example APIs could have body but user just want to ignore it).

----------------------

What about?
```kotlin
return if (response.isSuccessful) {
    if (successBodyType == Unit::class.java) {
        @Suppress("UNCHECKED_CAST")
        NetworkResponse.Success(Unit, headers) as NetworkResponse<S, E>
    } else {
        NetworkResponse.Success(body, headers)
    }
} else {
        val networkResponse: NetworkResponse<S, E> = try {
        val convertedBody = errorConverter.convert(errorBody)
        NetworkResponse.ServerError(convertedBody, code, headers)
    } catch (ex: Exception) {
        NetworkResponse.UnknownError(ex)
    }
    networkResponse
}
```

**Comment by haroldadmin** [comment]
> Ignored body on success may occur in REST requests, just relying on the status code. However, I would still need to parse unsuccessful responses (In that example APIs with status code 2xx could have body but user just want to ignore it).

I am having trouble understanding this use case. If a response is successful and you want the body to be ignored (even if it is non-empty) then you can just not consume it. If you want to indicate that this response's body will be ignored, then you can change the success type to be `Any`.

```kotlin
@GET("/something")
suspend fun getSomething(): NetworkResponse<Any, ErrorType>
```

I think I might be missing something. Can you give me an example of this flow so that I can understand it better?

**Comment by lucianocn** [comment]
My bad, what you said makes sense. So PR #14 for me is approved 👍

**Comment by haroldadmin** [comment]
@lucianocn I will add a status code field to the `NetworkResponse.Success` class. Maybe that would help in handling cases like the one you mention. 

This field should have been there from the beginning, I guess.

**Comment by lucianocn** [comment]
@haroldadmin that would be great!! 👍

**Comment by haroldadmin** [comment]
Closed via #14
